base/pkg/entry.jl

```diff
@@ -295,18 +295,42 @@ function pin(pkg::AbstractString, head::AbstractString)
     should_resolve = true
     with(GitRepo, pkg) do repo
         id = if isempty(head) # get HEAD commit
-            LibGit2.head_oid(repo)
-        else
             # no need to resolve, branch will be from HEAD
             should_resolve = false
+            LibGit2.head_oid(repo)
```

Comment:
`resolve` was being called when keeping the head fixed and not called when changing it, rather than the other way around.

The behaviour after the fix is the following (notice the info messages which tell you whether `resolve` was invoked or not):

```
julia> Pkg.pin("Example") # stay on head, resolve not invoked 
INFO: Creating Example branch pinned.526e34fd.tmp

julia> Pkg.pin("Example", v"0.4.0") # resolve invoked
INFO: Creating Example branch pinned.b1990792.tmp
INFO: No packages to install, update or remove

julia> Pkg.pin("Example", v"0.4.0") # no changes, resolve not invoked
INFO: Package Example is already pinned to the selected commit

julia> Pkg.pin("Example", v"0.4.1") # change commit, resolve invoked
INFO: Package Example: checking out existing branch pinned.526e34fd.tmp
INFO: No packages to install, update or remove

julia> Pkg.pin("Example") # no changes, resolve not invoked
INFO: Package Example is already pinned

julia> Pkg.free("Example") # free always calls resolve (may be changed, but it's not in this PR)
INFO: Freeing Example
INFO: No packages to install, update or remove
```